pkg/ipcache/ipcache.go

```diff
@@ -137,9 +137,11 @@ func (ipc *IPCache) SetListeners(listeners []IPIdentityMappingListener) {
 func (ipc *IPCache) AddListener(listener IPIdentityMappingListener) {
 	ipc.mutex.Lock()
 	ipc.listeners = append(ipc.listeners, listener)
+	ipc.mutex.Unlock()
+	ipc.mutex.RLock()
 	// Initialize new listener with the current mappings
 	ipc.DumpToListenerLocked(listener)
-	ipc.mutex.Unlock()
+	ipc.mutex.RUnlock()
```

Comment:
Releasing the lock in the middle can cause a race with an Upsert or Delete, which will result in to two separate change notifications for the same identity change, first from the Upsert and then from the Dump, or a Delete notification for an entry not yet dumped to the new listener. I checked the Envoy listener and it seems to handle both cases gracefully, so no problem here.

Reply:
@jrajahalme I can do a semaphore with weight 2, would be that be better?

```go
	// acquire 2
	ipc.listeners = append(ipc.listeners, listener)
	// release 1
	// Initialize new listener with the current mappings
	ipc.DumpToListenerLocked(listener)
	// release 1
```

all other Lock() calls would be replaced with `acquire 2` and all other RLocks would be replaced with `acquire 1`

Reply:
based on offline discussion a semaphore was implemented instead.